src/domFacade/DomFacade.ts

```diff
@@ -1,135 +1,485 @@
+					if (nextSibling && nextSibling.nodeType !== NODE_TYPES.DOCUMENT_TYPE_NODE) {
+						break;
+					}
+				}
 			}
-			return node2 as ConcreteChildNode;
 		}
-		return null;
+
+		if (!nextSibling) {
+			return null;
+		}
+
+		// Cache
+		let cacheForNode = this.pointerCache.get(nextSibling);
+		if (!cacheForNode) {
+			cacheForNode = new Map();
+			this.pointerCache.set(nextSibling, cacheForNode);
+		}
+		const unwrappedParentNode = this.getParentNode(pointer, bucket).unwrap();
+		let nextSiblingPointer = cacheForNode.get(unwrappedParentNode) as ChildNodePointer;
+		if (!nextSiblingPointer) {
+			let newGraftAncestor: GraftPoint = null;
+			if (graftAncestor) {
+				newGraftAncestor = {
+					graftAncestor: graftAncestor.graftAncestor,
+					offset: nextSiblingIndex,
+					parent: graftAncestor.parent
+				};
+			}
+			nextSiblingPointer = new ChildNodePointer(nextSibling, newGraftAncestor);
+			cacheForNode.set(unwrappedParentNode, nextSiblingPointer);
+		}
+		return nextSiblingPointer;
+	}
+
+	public getNodeName(pointer: ElementNodePointer | AttributeNodePointer): string {
+		return pointer.unwrap().nodeName;
+	}
+
+	public getNodeType(pointer: NodePointer): NODE_TYPES {
+		return pointer.unwrap().nodeType;
 	}
 
 	public getParentNode(
-		node: ConcreteChildNode,
+		pointer: ChildNodePointer,
 		bucket: string | null = null
-	): ConcreteParentNode {
-		return this._domFacade['getParentNode'](node, bucket) as ConcreteParentNode;
+	): ParentNodePointer {
+		const unwrappedChildNode = pointer.unwrap();
+		const graftAncestor = pointer.getGraftAncestor();
+		let unwrappedParentNode;
+		let newGraftAncestor;
+		let unwrappedGrandParentNode;
+
+		if (!graftAncestor) {
+			// check dom, or null if tinyNode
+			if (pointer.isTinyNode()) {
+				return null;
+			}
+			unwrappedParentNode = this._domFacade['getParentNode'](
+				unwrappedChildNode,
+				bucket
+			) as ConcreteParentNode;
+			unwrappedGrandParentNode =
+				unwrappedParentNode &&
+				(this._domFacade['getParentNode'](
+					unwrappedParentNode,
+					bucket
+				) as ConcreteParentNode);
+			newGraftAncestor = null;
+		} else if (
+			// check the child at graftAncestor is that node
+			(typeof graftAncestor.offset === 'number' &&
+				unwrappedChildNode ===
+					(graftAncestor.parent as TinyParentNode).childNodes[graftAncestor.offset]) ||
+			(typeof graftAncestor.offset === 'string' &&
+				unwrappedChildNode ===
+					(graftAncestor.parent as TinyElementNode).attributes.find(
+						e => graftAncestor.offset === e.nodeName
+					))
+		) {
+			unwrappedParentNode = graftAncestor.parent as TinyParentNode | ConcreteParentNode;
+			newGraftAncestor = graftAncestor.graftAncestor;
+			unwrappedGrandParentNode = newGraftAncestor ? newGraftAncestor.parent : null;
+		} else {
+			// if not go to the dom
+			if (pointer.isTinyNode()) {
+				// if you have a tinyNode, it should not be possible, throw error
+				throw new Error('Boom, there should not be tinyNode at this point');
```

Comment:
Do we have the need for the error i.e. have we seen bugs which encounter this? If so, consider the error message "Boom" does not provide much information as it's already a raised `Error`